src/_pytest/fixtures.py

```diff
@@ -1242,21 +1216,66 @@ def __call__(self, function: FixtureFunction) -> FixtureFunction:
                 pytrace=False,
             )
 
-        # Type ignored because https://github.com/python/mypy/issues/2087.
-        function._pytestfixturefunction = self  # type: ignore[attr-defined]
-        return function
+        return fixture_definition
+
+
+# TODO: paramspec/return type annotation tracking and storing
+class FixtureFunctionDefinition:
+    def __init__(
+        self,
+        *,
+        function: Callable[..., Any],
+        fixture_function_marker: FixtureFunctionMarker,
+        instance: object | None = None,
+        _ispytest: bool = False,
+    ) -> None:
+        check_ispytest(_ispytest)
+        self.name = fixture_function_marker.name or function.__name__
+        # In order to show the function that this fixture contains in messages.
+        # Set the __name__ to be same as the function __name__ or the given fixture name.
+        self.__name__ = self.name
+        self._fixture_function_marker = fixture_function_marker
+        self._fixture_function = function
+        self._instance = instance
```

Comment:
I wonder if it wouldn't be simpler to keep `self._fixture_function` bound to `instance` instead of keeping `_instance` and binding it lazily? Is there a reason why it needs to be lazy?

Reply:
I applied this and it's simpler. I don't think there's any reason to do it when we get the fixture.